File: 21. Merge Two Sorted Lists/solution_1.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

    def __str__(self):
        if self.next:
            return str(self.val) + ' -> ' + str(self.next)
        return str(self.val)


class Solution:
    def mergeTwoLists(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """

        def insert(pre, n1, n2):
            # if not pre:
            if not n1 and not n2:
                return
            elif not n1 or n2 and n1.val > n2.val:
                pre.next = n2
                insert(pre.next, n1, n2.next)
            else:
                pre.next = n1
                insert(pre.next, n1.next, n2)

        pre = ListNode(0)
        insert(pre, l1, l2)
        return pre.next


if __name__ == '__main__':
    # s = Solution()
    # l1 = ListNode(1)
    # l1.next = ListNode(2)
    # l1.next.next = ListNode(4)
    # l2 = ListNode(1)
    # l2.next = ListNode(3)
    # l2.next.next = ListNode(4)
    # # it = head
    # # for i in [2, 3, 4, 5]:
    # #     it.next = ListNode(i)
    # #     it = it.next
    # # s = Solution()
    # print(s.mergeTwoLists(l1, l2))

    s = Solution()
    l1 = ListNode(1)
    l2 = None
    print(s.mergeTwoLists(l2, l1))


# Runtime: 56 ms

File: 23. Merge k Sorted Lists/go/solution_1.go
package main

//Definition for singly-linked list.
type ListNode struct {
	Val int
	Next *ListNode
}


func mergeKLists(lists []*ListNode) *ListNode {
	var listsNotNil []*ListNode
	for _, v := range lists {
		if v != nil {
			listsNotNil = append(listsNotNil, v)
		}
	}
	var head = &ListNode{0,nil}
	merge(head, listsNotNil)
	return head.Next
}

func merge(chead *ListNode, lists []*ListNode){
	if len(lists) == 0 {
		return
	}
	minPos := 0
	for i:=1;i<len(lists);i++ {
		if lists[i].Val < lists[minPos].Val {
			minPos = i
		}
	}
	chead.Next = lists[minPos]
	if lists[minPos].Next == nil {
		lists = append(lists[:minPos],lists[minPos+1:]...)
	}else{
		lists[minPos] = lists[minPos].Next
	}
	merge(chead.Next, lists)
}

func main() {
	//l1 := generateListNode([]int{1,4,5})
	//l2 := generateListNode([]int{1,3,4})
	//l3 := generateListNode([]int{2,6})
	//merge := mergeKLists([]*ListNode{l1, l2, l3})
	l1 := generateListNode([]int{})
	merge := mergeKLists([]*ListNode{l1})
	printListNode(merge)
}

func generateListNode(n []int) *ListNode{
	var head *ListNode
	var pre *ListNode
	for i, v := range n {
		c := &ListNode{v,nil}
		if i != 0 {
			pre.Next = c
		}else{
			head = c
		}
		pre = c
	}
	return head
}

func printListNode(listNode *ListNode) {
	if listNode == nil{
		print("nil")
		return
	}
	print(listNode.Val)
	if listNode.Next != nil {
		print("->")
		printListNode(listNode.Next)
	}
}


//runtime: 168ms	Memory: 8.8MB
